server/src/test/java/org/elasticsearch/cluster/routing/allocation/allocator/AllocationActionMultiListenerTests.java

```diff
@@ -121,45 +123,60 @@ private static void awaitQuietly(CountDownLatch latch) {
+        final var expectedResponseHeader = randomAlphaOfLength(10);
 
-        assertThat(r1.get(), equalTo("root"));
-        assertThat(r2.get(), equalTo("root"));
-    }
+        var context = new ThreadContext(Settings.EMPTY);
+        var listener = new AllocationActionMultiListener<>(context);
+
+        context.putHeader(requestHeaderName, expectedRequestHeader);
+        context.addResponseHeader(responseHeaderName, expectedResponseHeader);
+
+        var isComplete = new AtomicBoolean();
+        try (var refs = new RefCountingRunnable(() -> assertTrue(isComplete.compareAndSet(false, true)))) {
+
+            List<Runnable> actions = new ArrayList<>();
+
+            for (int i = between(0, 5); i > 0; i--) {
+                var expectedVal = new Object();
+                var delayedListener = listener.delay(
+                    ActionListener.releaseAfter(ActionListener.wrap(Assert::fail).delegateFailure((l, val) -> {
+                        assertSame(expectedVal, val);
+                        assertEquals(expectedRequestHeader, context.getHeader(requestHeaderName));
+                        assertEquals(List.of(expectedResponseHeader), context.getResponseHeaders().get(responseHeaderName));
+                        context.addResponseHeader(responseHeaderName, randomAlphaOfLength(10));
+                    }), refs.acquire())
+                );
+                actions.add(() -> delayedListener.onResponse(expectedVal));
+            }
 
-    private static void executeInRandomOrder(ThreadContext context, List<Tuple<String, Runnable>> actions) {
-        for (var action : shuffledList(actions)) {
-            try (var ignored = context.stashContext()) {
-                context.putHeader("header", action.v1());
-                action.v2().run();
+            final var unexpectedResponseHeader = randomAlphaOfLength(10);
+            context.addResponseHeader(responseHeaderName, unexpectedResponseHeader);
+
+            actions.add(() -> listener.reroute().onResponse(null));
+
+            for (var action : shuffledList(actions)) {
+                try (var ignored = context.stashContext()) {
+                    final var localRequestHeader = randomAlphaOfLength(10);
+                    final var localResponseHeader = randomAlphaOfLength(10);
+                    context.putHeader(requestHeaderName, localRequestHeader);
+                    context.addResponseHeader(responseHeaderName, localResponseHeader);
+                    action.run();
+                    assertEquals(localRequestHeader, context.getHeader(requestHeaderName));
+                    assertEquals(List.of(localResponseHeader), context.getResponseHeaders().get(responseHeaderName));
+                }
             }
+
+            assertEquals(
+                Set.of(expectedResponseHeader, unexpectedResponseHeader),
+                Set.copyOf(context.getResponseHeaders().get(responseHeaderName))
```

Comment:
NIT: Consider renaming `unexpectedResponseHeader` -> `additionalResponseHeader` as it is expected by this time